distributed off-policy algorithm

hnyu opened the pull request:
This PR adds the first version of distributed off-policy algorithm. The idea is to support unrollers (responsible for interacting with an environment and collecting data) and a trainer (responsible for training the policy) on different machines.

1. Unrollers and the trainer can have different ips or the same ip.
2. Unrollers will send collected data to the trainer, while the trainer will send the updated model params to the unrollers. The data collection and params update are async. Basically we can think of splitting a typical off-policy algorithm into two complementary parts (unroll and train) that run on different machines.
3. A new unroller can joint at any time of the training. Also an existing unroller can be killed at any time. 
4. The trainer supports mutli-gpu training with DDP. Each DDP rank has its own replay buffer whose data gets routed from the unrollers evenly. Also the jobs of param syncing with the unrollers are assigned to DDP ranks evenly.
5. Currently only supports envs with batch_size=1 for each unroller for simplifying data routing decisions.

`DistributedOffPolicyAlgorithm` technically is not a new algorithm. Instead, it's just a wrapper around any existing off-policy algorithm (e.g., SAC, DDPG, etc). The wrapper is mainly used for establishing the communication protocol between unrollers and trainer.

A diagram of the pipeline can be found at https://horizonrobotics.feishu.cn/wiki/BFPewZ1mzi4GkMkOgUXcSsHHnEc

I tested this pipeline on cart_pole as in `distributed_sac_cartpole_conf.py`:

Training (two DDP ranks):
```bash
python -m alf.bin.train --root_dir /tmp/trainer --conf  distributed_sac_cartpole_conf.py --as_remote_trainer --conf_param="make_ddp_performer.find_unused_parameters=1" --distributed multi-gpu --conf_param="DistributedTrainer.push_params_every_n_iters=10"
```

Unrolling (in another terminal):
```bash
python -m alf.bin.train --root_dir /tmp/unroller --conf distributed_sac_cartpole_conf.py --as_remote_unroller
```

A second unroller (in a third terminal) 
```python
python -m alf.bin.train --root_dir /tmp/unroller1 --conf distributed_sac_cartpole_conf.py --as_remote_unroller
```

![image](https://github.com/user-attachments/assets/08fc1018-fb4d-4456-9879-e3f7b7ec7a21)


TODOs in the future:
- [ ] loading/writing persistent replay buffer on disk
- [ ] support env with batch_size>1 for the unroller

hnyu committed:
distributed off policy algorithm

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,541 @@
+        Because the unroller never creates an online replay buffer, we need this
+        hacked flag for to perform training in ``_train_iter_off_policy()``.
+        """
+        return True
```
Instead of overriding has_offline() and train_from_replay_buffer(), I feel just override train_iter() is easier to understand.

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,541 @@
+    while True:
+        try:
+            message = socket.recv_string(flags=zmq.NOBLOCK)
+            if message == "unroller: update":
```
It's better to use a global variable to store the message name, so that it can be shared at difference code locations.

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,541 @@
+            # Lock the params to avoid sending them to the unroller while being updated.
+            # We only need to lock for the main DDP rank because the other DDP ranks
+            # will sync with it.
+            with self._params_lock:
```
This is not reliable. There are other things besides parameters which can be changed during model training (e.g. various buffers). Instead of allowing sending parameters at anytime, it is more reliable to only send at the end of an iteration.

hnyu replied:
> Instead of allowing sending parameters at anytime, it is more reliable to only send at the end of an iteration.

I tried this method in the first place. But there is something tricky here. Imagine that rank 0 is responsible for sending the params to the unroller at an end of some training iter. When some other rank gets stuck due to `replay_buffer_not_ready` or `utd_exceeded`, it won't be able to perform backward. So rank 0 might also get stuck in DDP backward sync. If at this moment the unroller requests the param update, it will get stuck waiting for the reply from rank 0. Now if the unroller gets stuck, other ranks won't get more unroll data, and they won't get out of `replay_buffer_not_ready` or `utd_exceeded` loop. So the entire system have a deadlock. 

I haven't figured out a way how to solve this issue except using a separate thread to respond to the unroller.

hnyu replied:
> This is not reliable. There are other things besides parameters which can be changed during model training (e.g. various buffers). Instead of allowing sending parameters at anytime, it is more reliable to only send at the end of an iteration.

Now changed to trainer pushing params to unrollers and each unroller has a thread to receive params, without blocking unroll().

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,541 @@
+
+from typing import Callable
+import threading
+import multiprocessing as mp
```
Should be torch.multiprocessing?

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,541 @@
+
+    @property
+    def utd(self):
+        with self._total_exps_received.get_lock():
```
the total number of experiences can be obtained from `replay_buffer.get_current_position().sum()`

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,541 @@
+        socket.close()
+        cxt.term()
+
+    @property
```
I feel @property should be used only for some very simple stuff.

hnyu committed:
address comments

hnyu committed:
support multiple unrollers

hnyu commented:
Added several new majors changes:
1. Trainer pushes params at some interval, so there is no race condition between param update and param sending. Unroller uses a thread to receive the param, without blocking unroll().
2. Assuming only one trainer machine but multiple unroller machines, now we specify the trainer ip instead of unroller ip.
3. Support multiple unrollers. Because two unrollers might want to send exp data to the same DDP rank at the same time, each DDP rank maintains a table of unrollers' exp data. Only when one unroller finishes the current episode, its data is stored into the replay buffer. 

I have updated the PR description to include the new changes.

hnyu committed:
handle the case where zmq is not installed

hnyu committed:
handle dead unrollers

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,658 @@
+            return 0
+        return self._total_updates / total_exps
+
+    def _train_iter_off_policy(self):
```
Can just override train_iter() since train_iter is documented in algorithm_interface.py as part of the interface.

hnyu committed:
use SharedMemory for unroller params pulling

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,676 @@
+        size = len(buffer.getvalue())
+        # Create a shared memory object to store the new params
+        # The first char indicates whether the params have been updated
+        self._params = SharedMemory(create=True, size=size + 1, name='params')
```
It's better to make a more unique name to void conflict.

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,676 @@
+        # Check if the params have been updated
+        if bytes(self._params.buf[:1]) == b'1':
+            params = bytes(self._params.buf[1:])
+            buffer = io.BytesIO(params)
```
`bytes()` will copy the memory. Use `io.BytesIO(self._params.buf[1:]` to avoid copy.

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,676 @@
+            # We might only update part of the params
+            self._core_alg.load_state_dict(state_dict, strict=False)
+            logging.debug("Params updated from the trainer.")
+            self._params.buf[:1] = b'0'
```
Can use `self._params.buf[:1]=b'\0'` here, and use `if self._params.buf[0]==1` at line 647

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,676 @@
+    def _create_pull_params_subprocess(self):
+        # Compute the total size of the params
+        buffer = io.BytesIO()
+        torch.save(self._opt_free_state_dict(), buffer)
```
Does `torch.save` guarantee same size every time?

hnyu replied:
> Does `torch.save` guarantee same size every time?

Without optimizer state, I think so.

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,676 @@
+            # Sending params to all the connected unrollers.
+            dead_unrollers = []
+            for unroller_id in self._connected_unrollers:
+                if not self._send_params_to_unroller(unroller_id):
```
Distribute sending among workers: `if urroller_id % num_workers == ddp_rank`

hnyu replied:
> Distribute sending among workers: `if urroller_id % num_workers == ddp_rank`

We will detect if an unroller dies in `self._send_params_to_unroller`. If yes, the trainer should no longer send params to it. So It's relatively easier to manage this information in just one rank.

hnyu replied:
> Distribute sending among workers: `if urroller_id % num_workers == ddp_rank`

Updated. But this minor feature actually requires a significant code change as I need to let each unroller know which DDP rank (port) to connect to for params syncing. This port number is transmitted from the DDP rank during registration.

emailweixu replied:
record the time spent on sending parameter using `with record_time()` outside of the `if` statement

hnyu replied:
> record the time spent on sending parameter using `with record_time()` outside of the `if` statement

Added

hnyu committed:
address comments

hnyu committed:
remove the print()

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,705 @@
+    """A helper function for creating a ZMQ socket.
+
+    Args:
+        type: type of the socket.
```
Comment what are the valid values for `type`

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,705 @@
+        self.observe_for_replay = self._observe_for_replay
+
+        self._params_socket, _ = create_zmq_socket(
+            zmq.ROUTER, '*', self._port + _params_port_offset + self._ddp_rank)
```
This port is different from the comment at line 327

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,705 @@
+        return self._total_updates / total_exps
+
+    def _train_iter_off_policy(self):
+        if not self._daemons_started:
```
For continual training from an existing checkpoint, the checkpoint is loaded after the first `train_iter()` call in policy trainer. See `policy_trainer.Trainer._restore_checkpoint()`

hnyu replied:
> For continual training from an existing checkpoint, the checkpoint is loaded after the first `train_iter()` call in policy trainer. See `policy_trainer.Trainer._restore_checkpoint()`

Updated to create the thread and subprocess at iter=1.

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,705 @@
+        thread.start()
+
+    def _create_data_receiver_subprocess(self):
+        """Create a proc to receive experience data from unrollers.
```
proc -> process

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,705 @@
+            args=(self._replay_buffer, self._new_unroller_ips_and_ports,
+                  self._ddp_rank),
+            daemon=True)
+        process.start()
```
It seems that the subprocess will keep running even after the main process is terminated?
Need to have a mechanism for terminating it.

hnyu replied:
> It seems that the subprocess will keep running even after the main process is terminated? Need to have a mechanism for terminating it.

The daemon subprocess will terminate whenever the main process exits?

emailweixu replied:
Both Google search and ChatGPT says the contrary.

hnyu replied:
> Both Google search and ChatGPT says the contrary.

It's easy to verify this:

```python
import time
import multiprocessing as mp


def worker():
    while True:
        print("worker")
        time.sleep(1)


proc = mp.Process(target=worker, daemon=True)
proc.start()

time.sleep(2.)
print("Main process terminates")
```

And I've been using this feature for quite a long time.

emailweixu replied:
You are right. It seems that there are two types of daemon processes.

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,705 @@
+            args=(self._shared_alg_params.name, self._id,
+                  self._params_socket_k),
+            daemon=True)
+        process.start()
```
need to have mechanism for terminating the subprocess after the main process ends

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -0,0 +1,705 @@
+                             params_socket_k: int):
+    """ Once new params arrive, we put it in the shared memory and mark updated.
+    Later after the current unroll finishes, the unroller can load the
+    new params.
```
what is params_socket_k?

hnyu replied:
> what is params_socket_k?

Updated the variable name and docstring.

hnyu committed:
address comments

hnyu committed:
info -> debug

hnyu committed:
add record_time for param sending

emailweixu commented on alf/algorithms/distributed_off_policy_algorithm.py:
```diff
@@ -528,8 +535,9 @@ def _train_iter_off_policy(self):
             for unroller_id in self._unrollers_to_update_params:
-                if not self._send_params_to_unroller(unroller_id):
-                    dead_unrollers.append(unroller_id)
+                with record_time("time/trainer_send_params_to_unroller"):
```
should be outside the "if" at line 532 so that it can correctly compute the average time spent for each iteration.

hnyu committed:
fix record_time
